java/src/main/java/ai/rapids/cudf/JCudfSerialization.java

```diff
@@ -0,0 +1,412 @@
+        long nullCount = nullCounts[column];
+        TimeUnit tu = timeUnits[column];
+        DeviceMemoryBuffer validity = null;
+        DeviceMemoryBuffer data = null;
+        HostMemoryBuffer offsets = null;
+        boolean columnSuccess = false;
+        try {
+          if (nullCount > 0) {
+            long len = padFor64bitAlignment(BitVectorHelper.getValidityLengthInBytes(numRows));
+            validity = combinedBuffer.slice(combinedBufferOffset, len);
+            combinedBufferOffset += len;
+          }
+
+          if (type == DType.STRING || type == DType.STRING_CATEGORY) {
+            long offsetsLen = padFor64bitAlignment((numRows + 1) * 4);
+            offsets = combinedBufferOnHost.slice(combinedBufferOffset, offsetsLen);
+            combinedBufferOffset += offsetsLen;
+            int startStringOffset = offsets.getInt(0);
+            int endStringOffset = offsets.getInt(numRows * 4);
+            long deviceDataLen = padFor64bitAlignment(endStringOffset - startStringOffset);
+
+            if (deviceDataLen == 0) {
+              // The vector is full of null strings, this is a rare corner case, but here is the
+              // simplest place to work around it
+              data = DeviceMemoryBuffer.allocate(1);
+            } else {
+              data = combinedBuffer.slice(combinedBufferOffset, deviceDataLen);
+            }
+            combinedBufferOffset += deviceDataLen;
+          } else {
+            long deviceDataLen = padFor64bitAlignment(type.sizeInBytes * numRows);
+            data = combinedBuffer.slice(combinedBufferOffset, deviceDataLen);
+            combinedBufferOffset += deviceDataLen;
+          }
+          vectors[column] = new ColumnVector(type, tu, numRows, nullCount, data, validity, offsets, true);
+          columnSuccess = true;
+        } finally {
+          if (!columnSuccess) {
+            if (validity != null) {
+              validity.close();
+            }
+
+            if (data != null) {
+              data.close();
+            }
+
+            if(offsets != null) {
+              offsets.close();
+            }
+          }
+        }
+      }
+      tableSuccess = true;
+      return vectors;
+    } finally {
+      if (!tableSuccess) {
+        for (ColumnVector cv : vectors) {
+          if (cv != null) {
+            cv.close();
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Read a serialize table from teh given InputStream.
+   * @param in the stream to read the table data from.
+   * @return the deserialized table in device memory.
+   */
+  public static Table readTableFrom(InputStream in) throws IOException {
+    DataInputStream din;
+    if (in instanceof DataInputStream) {
+      din = (DataInputStream) in;
+    } else {
+      din = new DataInputStream(in);
+    }
+
+    int num = din.readInt();
+    if (num != SER_FORMAT_MAGIC_NUMBER) {
+      throw new IllegalStateException("THIS DOES NOT LOOK LIKE CUDF SERIALIZED DATA");
```

Comment:
Would be nice to show what was expected and what was received to aid in debugging.